core/App/ffi_operation.cpp

```diff
@@ -1535,4 +1535,82 @@ extern "C"
         return EH_OK;
     }
 
+    /**
+     * @brief Generate Hmac (SHA-256) with given apikey(encrypted), cmk and payload
+     * @param payload : Pass in the cmk, apikey(encrypted) and payload in the form of JSON string
+                {
+                    cmk : a base64 string,
+                    apikey: a base64 string
+                    payload : a string,
+                }
+     * @return [string] json string
+                {
+                    code: int,
+                    message: string,
+                    result: {
+                        hmac: string,
+                    }
+                }
+     */
+    uint32_t ffi_generateHmac(JsonObj payloadJson, char *respJson)
+    {
+        ehsm_status_t ret;
+        // input params
+        ehsm_keyblob_t *cmk = NULL;
+        ehsm_data_t *apikey = NULL;
+        ehsm_data_t *payload = NULL;
+        // immediate vars
+        size_t payload_size;
+        std::string payload_str;
+        // output params
+        ehsm_data_t *hmac = NULL;
+        std::string hmac_str;
+        RetJsonObj retJsonObj;
+
+        // 0. prepare data
+        JSON2STRUCT(payloadJson, cmk);
+        JSON2STRUCT(payloadJson, apikey);
+
+        payload_str = payloadJson.readData_string("payload");
+        payload_size = payload_str.size();
+        payload = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(payload_size));
+        if (payload == NULL)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_BAD_REQUEST);
+            retJsonObj.setMessage("The payload's length is invalid.");
+            goto out;
+        }
+        payload->datalen = payload_size;
+        memcpy_s(payload->data, payload_size, (uint8_t *)payload_str.data(), payload_size);
```

Comment:
payload_str not do base64 decode, because the user side also do the  hmac with b64 string, right?  you can add some comments here, because it's some  difference with the other functions

Reply:
Thanks for reviewing!

According to the original implementation of `gen_hmac()`, the user side does the hmac with pure string.

The string used for hmac is constructed by concatenating `sign_params` to format like `key1=value1&key2=value2` . The implementation is in [ehsm/ehsm_kms_server/function.js](https://github.com/intel/ehsm/blob/c9a8ea0248ad2457cd2d24fbb8c1c725566036ee/ehsm_kms_service/function.js#L357).